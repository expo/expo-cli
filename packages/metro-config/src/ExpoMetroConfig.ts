import path from 'path';

import { ProjectTarget, getConfig, getDefaultTarget, resolveModule } from '@expo/config';
import { getBareExtensions, getManagedExtensions } from '@expo/config/paths';
import { Reporter } from 'metro';
import { ConfigT, InputConfigT, loadConfig } from 'metro-config';

const INTERNAL_CALLSITES_REGEX = new RegExp(
  [
    '/Libraries/Renderer/implementations/.+\\.js$',
    '/Libraries/BatchedBridge/MessageQueue\\.js$',
    '/Libraries/YellowBox/.+\\.js$',
    '/Libraries/LogBox/.+\\.js$',
    '/Libraries/Core/Timers/.+\\.js$',
    '/node_modules/react-devtools-core/.+\\.js$',
    '/node_modules/react-refresh/.+\\.js$',
    '/node_modules/scheduler/.+\\.js$',
  ].join('|')
);

export interface DefaultConfigOptions {
  target?: ProjectTarget;
}

export function getDefaultConfig(
  projectRoot: string,
  options: DefaultConfigOptions = {}
): InputConfigT {
  const { exp } = getConfig(projectRoot, { skipSDKVersionRequirement: true });
  const reactNativePath = resolveModule('react-native', projectRoot, exp);

  const target = options.target ?? process.env.EXPO_TARGET ?? getDefaultTarget(projectRoot);
  if (!(target === 'managed' || target === 'bare')) {
    throw new Error(
      `Invalid target: '${target}'. Debug info: \n${JSON.stringify(
        {
          'options.target': options.target,
          EXPO_TARGET: process.env.EXPO_TARGET,
          default: getDefaultTarget(projectRoot),
        },
        null,
        2
      )}`
    );
  }
  const sourceExtsConfig = { isTS: true, isReact: true, isModern: false };
  const sourceExts =
    target === 'bare'
      ? getBareExtensions([], sourceExtsConfig)
      : getManagedExtensions([], sourceExtsConfig);

  return {
    resolver: {
      resolverMainFields: ['react-native', 'browser', 'main'],
      platforms: ['ios', 'android', 'native'],
      sourceExts,
    },
    serializer: {
      getModulesRunBeforeMainModule: () => [
        require.resolve(path.join(reactNativePath, 'Libraries/Core/InitializeCore')),
        // TODO: Bacon: load Expo side-effects
      ],
      getPolyfills: () => require(path.join(reactNativePath, 'rn-get-polyfills'))(),
    },
    symbolicator: {
      customizeFrame: (frame: { file: string | null }) => {
        const collapse = Boolean(frame.file && INTERNAL_CALLSITES_REGEX.test(frame.file));
        return { collapse };
      },
    },
    transformer: {
      babelTransformerPath: require.resolve('../transformer'),
      // TODO: Bacon: Add path for web platform
      assetRegistryPath: path.join(reactNativePath, 'Libraries/Image/AssetRegistry'),
      assetPlugins: [resolveModule('expo/tools/hashAssetFiles', projectRoot, exp)],
    },
  };
}

export interface LoadOptions {
  config?: string;
  maxWorkers?: number;
  port?: number;
  reporter?: Reporter;
  resetCache?: boolean;
  target?: ProjectTarget;
}

export async function loadAsync(
  projectRoot: string,
  { reporter, target, ...metroOptions }: LoadOptions = {}
): Promise<ConfigT> {
  let defaultConfig = getDefaultConfig(projectRoot, { target });
  if (reporter) {
    defaultConfig = { ...defaultConfig, reporter };
  }
  return await loadConfig({ cwd: projectRoot, projectRoot, ...metroOptions }, defaultConfig);
}
