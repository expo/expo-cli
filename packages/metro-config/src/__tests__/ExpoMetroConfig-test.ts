import path from 'path';

import { getDefaultConfig, loadAsync } from '../ExpoMetroConfig';

const projectRoot = path.join(__dirname, '__fixtures__', 'hello-world');

describe('getDefaultConfig', () => {
  const propertyMatchers = {
    transformer: {
      assetPlugins: expect.arrayContaining([expect.stringContaining('hashAssetFiles')]),
      assetRegistryPath: expect.stringContaining('AssetRegistry'),
      babelTransformerPath: expect.stringContaining('transformer.js'),
    },
  };

  it('loads default configuration', () => {
    expect(getDefaultConfig(projectRoot)).toMatchSnapshot(propertyMatchers);
  });

  it('loads default configuration for bare apps', () => {
    expect(getDefaultConfig(projectRoot, { target: 'bare' })).toMatchSnapshot(propertyMatchers);
  });

  it('complains about an invalid target setting', () => {
    process.env.EXPO_TARGET = 'bare';
    expect(() =>
      // @ts-ignore incorrect `target` value passed on purpose
      getDefaultConfig(projectRoot, { target: 'blooper' })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('loadAsync', () => {
  it('adds runtime options to the default configuration', async () => {
    const options = {
      maxWorkers: 10,
      resetCache: true,
      reporter: { update() {} },
      sourceExts: ['yml', 'toml', 'json'],
    };
    const config = await loadAsync(projectRoot, options);
    expect(config).toMatchObject({
      maxWorkers: options.maxWorkers,
      resetCache: options.resetCache,
      reporter: options.reporter,
      resolver: {
        sourceExts: options.sourceExts,
      },
    });
  });
});
