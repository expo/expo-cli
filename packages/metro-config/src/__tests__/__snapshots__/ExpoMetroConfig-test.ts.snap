// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDefaultConfig complains about an invalid target setting 1`] = `
"Invalid target: 'blooper'. Debug info: 
{
  \\"options.target\\": \\"blooper\\",
  \\"EXPO_TARGET\\": \\"bare\\",
  \\"default\\": \\"managed\\"
}"
`;
